# Digital Connectivity and Internet Adoption in Africa (2010–2024)

In [218]:
import pandas as pd

df = pd.read_csv("WDICSV.csv")

In [219]:
selected_indicators = [
    "IT.NET.USER.ZS",
    "IT.CEL.SETS.P2",
    "IT.NET.BBND.P2"
]

tech_data = df[df["Indicator Code"].isin(selected_indicators)].copy()

tech_data.shape

(795, 70)

In [220]:
tech_data[["Country Name", "Country Code", "Indicator Name", "Indicator Code"]].head(20)

,Country Name,Country Code,Indicator Name,Indicator Code
489,Africa Eastern and Southern,AFE,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2
662,Africa Eastern and Southern,AFE,Individuals using the Internet (% of population),IT.NET.USER.ZS
834,Africa Eastern and Southern,AFE,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2
1987,Africa Western and Central,AFW,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2
2160,Africa Western and Central,AFW,Individuals using the Internet (% of population),IT.NET.USER.ZS
2332,Africa Western and Central,AFW,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2
3485,Arab World,ARB,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2
3658,Arab World,ARB,Individuals using the Internet (% of population),IT.NET.USER.ZS
3830,Arab World,ARB,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2
4983,Caribbean small states,CSS,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2


In [221]:
tech_data["Indicator Name"].value_counts()

Indicator Name
Fixed broadband subscriptions (per 100 people)      265
Individuals using the Internet (% of population)    265
Mobile cellular subscriptions (per 100 people)      265
Name: count, dtype: int64

In [222]:
tech_data["Country Name"].nunique()

265

In [223]:
tech_data.columns.tolist()

['Country Name',
 'Country Code',
 'Indicator Name',
 'Indicator Code',
 '1960',
 '1961',
 '1962',
 '1963',
 '1964',
 '1965',
 '1966',
 '1967',
 '1968',
 '1969',
 '1970',
 '1971',
 '1972',
 '1973',
 '1974',
 '1975',
 '1976',
 '1977',
 '1978',
 '1979',
 '1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013',
 '2014',
 '2015',
 '2016',
 '2017',
 '2018',
 '2019',
 '2020',
 '2021',
 '2022',
 '2023',
 '2024',
 '2025']

In [224]:
tech_data[["Country Name", "Country Code"]].drop_duplicates().to_string(index=False)

"                                                             Country Name Country Code\n                                              Africa Eastern and Southern          AFE\n                                               Africa Western and Central          AFW\n                                                               Arab World          ARB\n                                                   Caribbean small states          CSS\n                                           Central Europe and the Baltics          CEB\n                                               Early-demographic dividend          EAR\n                                                      East Asia & Pacific          EAS\n                              East Asia & Pacific (excluding high income)          EAP\n                               East Asia & Pacific (IDA & IBRD countries)          TEA\n                                                                Euro area          EMU\n                               

In [225]:
for indicator in selected_indicators:
    temp = tech_data[tech_data["Indicator Code"] == indicator]
    
    year_columns = [col for col in temp.columns if col.isdigit()]
    
    available_years = [
        int(year) for year in year_columns
        if temp[year].notna().any()
    ]
    
    print(
        temp["Indicator Name"].iloc[0],
        "→",
        min(available_years),
        "to",
        max(available_years)
    )

Individuals using the Internet (% of population) → 1990 to 2025
Mobile cellular subscriptions (per 100 people) → 1960 to 2025
Fixed broadband subscriptions (per 100 people) → 1998 to 2025


In [226]:
africa_codes = [
    "DZA", "AGO", "BEN", "BWA", "BFA", "BDI", "CPV", "CMR",
    "CAF", "TCD", "COM", "COD", "COG", "CIV", "DJI", "EGY",
    "GNQ", "ERI", "SWZ", "ETH", "GAB", "GMB", "GHA", "GIN",
    "GNB", "KEN", "LSO", "LBR", "LBY", "MDG", "MWI", "MLI",
    "MRT", "MUS", "MAR", "MOZ", "NAM", "NER", "NGA", "RWA",
    "STP", "SEN", "SYC", "SLE", "SOM", "ZAF", "SSD", "SDN",
    "TZA", "TGO", "TUN", "UGA", "ZMB", "ZWE"
]

africa_data = tech_data[
    tech_data["Country Code"].isin(africa_codes)
].copy()

print("Rows:", africa_data.shape[0])
print("Countries:", africa_data["Country Name"].nunique())

africa_data[["Country Name", "Country Code"]].drop_duplicates()

Rows: 162
Countries: 54


,Country Name,Country Code
75389,Algeria,DZA
79883,Angola,AGO
102353,Benin,BEN
109843,Botswana,BWA
117333,Burkina Faso,BFA
118831,Burundi,BDI
120329,Cabo Verde,CPV
123325,Cameroon,CMR
127819,Central African Republic,CAF
129317,Chad,TCD


In [227]:
year_columns = [str(year) for year in range(1998, 2026)]

missing_values = africa_data[year_columns].isna().sum()

missing_values

1998     57
1999     56
2000     54
2001     53
2002     45
2003     42
2004     34
2005     25
2006     27
2007     16
2008     12
2009     12
2010      7
2011      7
2012      3
2013      3
2014      4
2015      2
2016      2
2017      5
2018     17
2019      2
2020      7
2021      5
2022      6
2023     23
2024     35
2025    162
dtype: int64

In [228]:
missing_by_indicator = (
    africa_data
    .groupby("Indicator Name")[year_columns]
    .apply(lambda x: x.isna().sum().sum())
)

missing_by_indicator

Indicator Name
Fixed broadband subscriptions (per 100 people)      514
Individuals using the Internet (% of population)    110
Mobile cellular subscriptions (per 100 people)       99
dtype: int64

In [229]:
missing_by_year_indicator = (
    africa_data
    .groupby("Indicator Name")[year_columns]
    .apply(lambda x: x.isna().sum())
)

missing_by_year_indicator

,1998,1999,2000,2001,2002,2003,2004,2005,2006,2007,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
Indicator Name,,,,,,,,,,,,,,,,,,,,,
Fixed broadband subscriptions (per 100 people),54,54,52,51,43,39,30,20,20,14,...,2,3,9,1,3,1,2,10,17,54
Individuals using the Internet (% of population),2,1,1,1,1,2,2,3,4,1,...,0,2,3,1,3,4,4,5,3,54
Mobile cellular subscriptions (per 100 people),1,1,1,1,1,1,2,2,3,1,...,0,0,5,0,1,0,0,8,15,54


In [230]:
africa_data.groupby("Indicator Name")["2025"].apply(lambda x: x.notna().sum())

Indicator Name
Fixed broadband subscriptions (per 100 people)      0
Individuals using the Internet (% of population)    0
Mobile cellular subscriptions (per 100 people)      0
Name: 2025, dtype: int64

In [231]:
completeness = (
    africa_data
    .groupby("Indicator Name")[year_columns]
    .apply(lambda x: x.notna().mean() * 100)
)

completeness

,1998,1999,2000,2001,2002,2003,2004,2005,2006,2007,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
Indicator Name,,,,,,,,,,,,,,,,,,,,,
Fixed broadband subscriptions (per 100 people),0.000000,0.000000,3.703704,5.555556,20.370370,27.777778,44.444444,62.962963,62.962963,74.074074,...,96.296296,94.444444,83.333333,98.148148,94.444444,98.148148,96.296296,81.481481,68.518519,0.0
Individuals using the Internet (% of population),96.296296,98.148148,98.148148,98.148148,98.148148,96.296296,96.296296,94.444444,92.592593,98.148148,...,100.000000,96.296296,94.444444,98.148148,94.444444,92.592593,92.592593,90.740741,94.444444,0.0
Mobile cellular subscriptions (per 100 people),98.148148,98.148148,98.148148,98.148148,98.148148,98.148148,96.296296,96.296296,94.444444,98.148148,...,100.000000,100.000000,90.740741,100.000000,98.148148,100.000000,100.000000,85.185185,72.222222,0.0


In [232]:
analysis_years = [str(year) for year in range(2010, 2026)]

for indicator in selected_indicators:
    temp = africa_data[
        africa_data["Indicator Code"] == indicator
    ][analysis_years]

    total_values = temp.size
    missing = temp.isna().sum().sum()
    completeness = (1 - missing / total_values) * 100

    print(
        indicator,
        "→ Missing:",
        missing,
        "| Completeness:",
        round(completeness, 2),
        "%"
    )

IT.NET.USER.ZS → Missing: 87 | Completeness: 89.93 %
IT.CEL.SETS.P2 → Missing: 83 | Completeness: 90.39 %
IT.NET.BBND.P2 → Missing: 120 | Completeness: 86.11 %


In [233]:
analysis_data = africa_data.copy()

analysis_years = [str(year) for year in range(2010, 2026)]

analysis_data = analysis_data[
    ["Country Name", "Country Code",
     "Indicator Name", "Indicator Code"] + analysis_years
].copy()

analysis_data.head()

,Country Name,Country Code,Indicator Name,Indicator Code,2010,2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
75389,Algeria,DZA,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2.486996,2.657621,3.067372,3.332104,4.080323,5.670601,7.000040,7.700458,7.215668,8.275266,8.606447,9.331842,10.351724,12.008684,13.317309,NaN
75562,Algeria,DZA,Individuals using the Internet (% of population),IT.NET.USER.ZS,12.500000,14.900000,18.200001,22.500000,29.500000,38.200001,42.945499,47.691101,49.038468,55.790699,63.472698,72.212402,74.781502,76.197601,77.417801,NaN
75734,Algeria,DZA,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2,90.582379,96.511295,99.685325,102.871008,110.440351,108.016371,115.154200,109.969863,110.938067,104.922068,103.436672,105.037093,107.787116,111.606236,115.457279,NaN
79883,Angola,AGO,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,0.064392,0.065310,0.081470,0.085158,0.323076,0.545394,0.290319,0.320554,0.350067,0.367770,0.363264,0.390925,0.386698,0.373669,0.362845,NaN
80056,Angola,AGO,Individuals using the Internet (% of population),IT.NET.USER.ZS,2.800000,4.700000,7.700000,13.000000,21.362341,22.000000,23.200001,26.000000,29.000000,32.129398,33.320900,34.556499,35.838001,37.885899,40.704800,NaN


In [234]:
duplicates = analysis_data.duplicated().sum()

print("Duplicate rows:", duplicates)

Duplicate rows: 0


In [235]:
analysis_data.dtypes

Country Name          str
Country Code          str
Indicator Name        str
Indicator Code        str
2010              float64
2011              float64
2012              float64
2013              float64
2014              float64
2015              float64
2016              float64
2017              float64
2018              float64
2019              float64
2020              float64
2021              float64
2022              float64
2023              float64
2024              float64
2025              float64
dtype: object

In [236]:
missing_2025 = (
    analysis_data
    .groupby("Indicator Code")["2025"]
    .apply(lambda x: x.isna().sum())
)

print(missing_2025)

Indicator Code
IT.CEL.SETS.P2    54
IT.NET.BBND.P2    54
IT.NET.USER.ZS    54
Name: 2025, dtype: int64


In [237]:
missing_2024 = (
    analysis_data
    .groupby("Indicator Code")["2024"]
    .apply(lambda x: x.isna().sum())
)

print(missing_2024)

Indicator Code
IT.CEL.SETS.P2    15
IT.NET.BBND.P2    17
IT.NET.USER.ZS     3
Name: 2024, dtype: int64


In [238]:
analysis_data[["Country Name", "Indicator Code", "2024", "2025"]].head(15)

,Country Name,Indicator Code,2024,2025
75389,Algeria,IT.NET.BBND.P2,13.317309,NaN
75562,Algeria,IT.NET.USER.ZS,77.417801,NaN
75734,Algeria,IT.CEL.SETS.P2,115.457279,NaN
79883,Angola,IT.NET.BBND.P2,0.362845,NaN
80056,Angola,IT.NET.USER.ZS,40.704800,NaN
80228,Angola,IT.CEL.SETS.P2,69.703947,NaN
102353,Benin,IT.NET.BBND.P2,0.277382,NaN
102526,Benin,IT.NET.USER.ZS,33.983799,NaN
102698,Benin,IT.CEL.SETS.P2,125.940874,NaN
109843,Botswana,IT.NET.BBND.P2,7.440962,NaN


In [239]:
print(analysis_data["2025"].notna().sum())
print(analysis_data["2025"].isna().sum())

0
162


In [240]:
africa_data[["Country Name", "Indicator Code", "2025"]].head(15)

,Country Name,Indicator Code,2025
75389,Algeria,IT.NET.BBND.P2,NaN
75562,Algeria,IT.NET.USER.ZS,NaN
75734,Algeria,IT.CEL.SETS.P2,NaN
79883,Angola,IT.NET.BBND.P2,NaN
80056,Angola,IT.NET.USER.ZS,NaN
80228,Angola,IT.CEL.SETS.P2,NaN
102353,Benin,IT.NET.BBND.P2,NaN
102526,Benin,IT.NET.USER.ZS,NaN
102698,Benin,IT.CEL.SETS.P2,NaN
109843,Botswana,IT.NET.BBND.P2,NaN


In [241]:
analysis_years = [str(year) for year in range(2010, 2025)]

missing_by_year = (
    analysis_data[analysis_years]
    .isna()
    .sum()
)

print(missing_by_year)

2010     7
2011     7
2012     3
2013     3
2014     4
2015     2
2016     2
2017     5
2018    17
2019     2
2020     7
2021     5
2022     6
2023    23
2024    35
dtype: int64


In [242]:
missing_by_indicator = (
    analysis_data
    .groupby("Indicator Code")[analysis_years]
    .apply(lambda x: x.isna().sum())
)

missing_by_indicator

,2010,2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024
Indicator Code,,,,,,,,,,,,,,,
IT.CEL.SETS.P2,0,0,0,0,0,0,0,0,5,0,1,0,0,8,15
IT.NET.BBND.P2,4,5,2,2,3,2,2,3,9,1,3,1,2,10,17
IT.NET.USER.ZS,3,2,1,1,1,0,0,2,3,1,3,4,4,5,3


In [243]:
clean_data = analysis_data.copy()

analysis_years = [str(year) for year in range(2010, 2025)]

for indicator in clean_data["Indicator Code"].unique():
    
    mask = clean_data["Indicator Code"] == indicator
    
    clean_data.loc[mask, analysis_years] = (
        clean_data.loc[mask, analysis_years]
        .interpolate(axis=1, limit_area="inside")
    )

print("Missing values after interpolation:")

print(
    clean_data[analysis_years]
    .isna()
    .sum()
)

Missing values after interpolation:
2010     7
2011     5
2012     2
2013     2
2014     2
2015     1
2016     1
2017     1
2018     2
2019     1
2020     2
2021     3
2022     4
2023    21
2024    35
dtype: int64


In [244]:
limit_area="inside"



In [245]:
remaining_missing = clean_data[analysis_years].isna().sum().sum()

print("Remaining missing values:", remaining_missing)

Remaining missing values: 89


In [246]:
clean_data.isna().sum()

Country Name        0
Country Code        0
Indicator Name      0
Indicator Code      0
2010                7
2011                5
2012                2
2013                2
2014                2
2015                1
2016                1
2017                1
2018                2
2019                1
2020                2
2021                3
2022                4
2023               21
2024               35
2025              162
dtype: int64

In [247]:
clean_data = clean_data.drop(columns=["2025"])

print(clean_data.columns)

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017', '2018',
       '2019', '2020', '2021', '2022', '2023', '2024'],
      dtype='str')


In [248]:
print(
    "Remaining missing values:",
    clean_data.isna().sum().sum()
)

Remaining missing values: 89


In [249]:
print("Number of countries:", clean_data["Country Name"].nunique())

print("\nNumber of indicators:", clean_data["Indicator Code"].nunique())

print("\nIndicators:")
print(clean_data[["Indicator Name", "Indicator Code"]].drop_duplicates())

Number of countries: 54

Number of indicators: 3

Indicators:
                                         Indicator Name  Indicator Code
75389    Fixed broadband subscriptions (per 100 people)  IT.NET.BBND.P2
75562  Individuals using the Internet (% of population)  IT.NET.USER.ZS
75734    Mobile cellular subscriptions (per 100 people)  IT.CEL.SETS.P2


In [250]:
long_data = clean_data.melt(
    id_vars=[
        "Country Name",
        "Country Code",
        "Indicator Name",
        "Indicator Code"
    ],
    value_vars=[str(year) for year in range(2010, 2025)],
    var_name="Year",
    value_name="Value"
)

long_data.head(10)

,Country Name,Country Code,Indicator Name,Indicator Code,Year,Value
0,Algeria,DZA,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2010,2.486996
1,Algeria,DZA,Individuals using the Internet (% of population),IT.NET.USER.ZS,2010,12.500000
2,Algeria,DZA,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2,2010,90.582379
3,Angola,AGO,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2010,0.064392
4,Angola,AGO,Individuals using the Internet (% of population),IT.NET.USER.ZS,2010,2.800000
5,Angola,AGO,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2,2010,40.366755
6,Benin,BEN,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2010,0.275928
7,Benin,BEN,Individuals using the Internet (% of population),IT.NET.USER.ZS,2010,3.130000
8,Benin,BEN,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2,2010,72.211539
9,Botswana,BWA,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2010,0.589146


In [251]:
long_data["Year"] = long_data["Year"].astype(int)

print("Rows:", len(long_data))
print("Columns:", long_data.columns.tolist())

Rows: 2430
Columns: ['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code', 'Year', 'Value']


In [252]:
indicator_names = {
    "IT.NET.USER.ZS": "Internet Users (%)",
    "IT.CEL.SETS.P2": "Mobile Subscriptions",
    "IT.NET.BBND.P2": "Fixed Broadband"
}

long_data["Indicator"] = long_data["Indicator Code"].map(indicator_names)

long_data.head(10)

,Country Name,Country Code,Indicator Name,Indicator Code,Year,Value,Indicator
0,Algeria,DZA,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2010,2.486996,Fixed Broadband
1,Algeria,DZA,Individuals using the Internet (% of population),IT.NET.USER.ZS,2010,12.500000,Internet Users (%)
2,Algeria,DZA,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2,2010,90.582379,Mobile Subscriptions
3,Angola,AGO,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2010,0.064392,Fixed Broadband
4,Angola,AGO,Individuals using the Internet (% of population),IT.NET.USER.ZS,2010,2.800000,Internet Users (%)
5,Angola,AGO,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2,2010,40.366755,Mobile Subscriptions
6,Benin,BEN,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2010,0.275928,Fixed Broadband
7,Benin,BEN,Individuals using the Internet (% of population),IT.NET.USER.ZS,2010,3.130000,Internet Users (%)
8,Benin,BEN,Mobile cellular subscriptions (per 100 people),IT.CEL.SETS.P2,2010,72.211539,Mobile Subscriptions
9,Botswana,BWA,Fixed broadband subscriptions (per 100 people),IT.NET.BBND.P2,2010,0.589146,Fixed Broadband


In [253]:


print(long_data["Indicator"].value_counts())

Indicator
Fixed Broadband         810
Internet Users (%)      810
Mobile Subscriptions    810
Name: count, dtype: int64


In [254]:
feature_data = long_data.pivot_table(
    index=["Country Name", "Country Code", "Year"],
    columns="Indicator",
    values="Value"
).reset_index()

feature_data["Internet-Mobile Gap"] = (
    feature_data["Mobile Subscriptions"]
    - feature_data["Internet Users (%)"]
)

print(feature_data.head(10))

Indicator Country Name Country Code  Year  Fixed Broadband  \
0              Algeria          DZA  2010         2.486996   
1              Algeria          DZA  2011         2.657621   
2              Algeria          DZA  2012         3.067372   
3              Algeria          DZA  2013         3.332104   
4              Algeria          DZA  2014         4.080323   
5              Algeria          DZA  2015         5.670601   
6              Algeria          DZA  2016         7.000040   
7              Algeria          DZA  2017         7.700458   
8              Algeria          DZA  2018         7.215668   
9              Algeria          DZA  2019         8.275266   

Indicator  Internet Users (%)  Mobile Subscriptions  Internet-Mobile Gap  
0                   12.500000             90.582379            78.082379  
1                   14.900000             96.511295            81.611295  
2                   18.200001             99.685325            81.485324  
3                

In [255]:
print(feature_data.columns)

Index(['Country Name', 'Country Code', 'Year', 'Fixed Broadband',
       'Internet Users (%)', 'Mobile Subscriptions', 'Internet-Mobile Gap'],
      dtype='str', name='Indicator')


In [256]:
feature_data["Internet-Broadband Gap"] = (
    feature_data["Internet Users (%)"]
    - feature_data["Fixed Broadband"]
)

print(
    feature_data[
        [
            "Country Name",
            "Year",
            "Internet Users (%)",
            "Fixed Broadband",
            "Internet-Broadband Gap"
        ]
    ].head(10)
)

Indicator Country Name  Year  Internet Users (%)  Fixed Broadband  \
0              Algeria  2010           12.500000         2.486996   
1              Algeria  2011           14.900000         2.657621   
2              Algeria  2012           18.200001         3.067372   
3              Algeria  2013           22.500000         3.332104   
4              Algeria  2014           29.500000         4.080323   
5              Algeria  2015           38.200001         5.670601   
6              Algeria  2016           42.945499         7.000040   
7              Algeria  2017           47.691101         7.700458   
8              Algeria  2018           49.038468         7.215668   
9              Algeria  2019           55.790699         8.275266   

Indicator  Internet-Broadband Gap  
0                       10.013004  
1                       12.242378  
2                       15.132629  
3                       19.167896  
4                       25.419677  
5                       

In [257]:
score_columns = [
    "Internet Users (%)",
    "Mobile Subscriptions",
    "Fixed Broadband"
]

for col in score_columns:
    min_value = feature_data[col].min()
    max_value = feature_data[col].max()

    feature_data[f"{col} Normalized"] = (
        (feature_data[col] - min_value)
        / (max_value - min_value)
    )

feature_data["Digital Connectivity Score"] = (
    feature_data["Internet Users (%) Normalized"]
    + feature_data["Mobile Subscriptions Normalized"]
    + feature_data["Fixed Broadband Normalized"]
) / 3

print(
    feature_data[
        [
            "Country Name",
            "Year",
            "Internet Users (%)",
            "Mobile Subscriptions",
            "Fixed Broadband",
            "Digital Connectivity Score"
        ]
    ].head(10)
)

Indicator Country Name  Year  Internet Users (%)  Mobile Subscriptions  \
0              Algeria  2010           12.500000             90.582379   
1              Algeria  2011           14.900000             96.511295   
2              Algeria  2012           18.200001             99.685325   
3              Algeria  2013           22.500000            102.871008   
4              Algeria  2014           29.500000            110.440351   
5              Algeria  2015           38.200001            108.016371   
6              Algeria  2016           42.945499            115.154200   
7              Algeria  2017           47.691101            109.969863   
8              Algeria  2018           49.038468            110.938067   
9              Algeria  2019           55.790699            104.922068   

Indicator  Fixed Broadband  Digital Connectivity Score  
0                 2.486996                    0.218320  
1                 2.657621                    0.239253  
2             

In [258]:
cleaning_summary = {
    "Dataset": "World Development Indicators (WDI)",
    "Theme": "Technology and Internet Adoption",
    "Countries": 54,
    "Years": "2010–2024",
    "Indicators": 3,
    "Missing Values": "Checked and handled where appropriate",
    "Duplicates": "Checked - 0 duplicate rows",
    "Data Format": "Converted from wide format to long format",
    "Standardization": "Year and indicator formats standardized",
    "Calculated Features": [
        "Internet-Mobile Gap",
        "Internet-Broadband Gap",
        "Digital Connectivity Score"
    ]
}

for key, value in cleaning_summary.items():
    print(f"{key}: {value}")

Dataset: World Development Indicators (WDI)
Theme: Technology and Internet Adoption
Countries: 54
Years: 2010–2024
Indicators: 3
Missing Values: Checked and handled where appropriate
Duplicates: Checked - 0 duplicate rows
Data Format: Converted from wide format to long format
Standardization: Year and indicator formats standardized
Calculated Features: ['Internet-Mobile Gap', 'Internet-Broadband Gap', 'Digital Connectivity Score']


In [259]:
feature_data.to_csv(
    "digital_connectivity_africa.csv",
    index=False
)

print("Power BI dataset saved successfully!")

Power BI dataset saved successfully!


### Data Analysis

1- Trend and Patterns
The analysis examines changes in digital connectivity across 54 African countries from 2010 to 2024. The main indicators analyzed are Internet Users (%), Mobile Subscriptions, and Fixed Broadband subscriptions.


In [260]:
trend_analysis = long_data.groupby(["Year", "Indicator"])["Value"].mean().reset_index()

trend_analysis

,Year,Indicator,Value
0,2010,Fixed Broadband,0.718634
1,2010,Internet Users (%),9.054510
2,2010,Mobile Subscriptions,54.740167
3,2011,Fixed Broadband,0.893669
4,2011,Internet Users (%),10.285833
5,2011,Mobile Subscriptions,62.467689
6,2012,Fixed Broadband,0.954845
7,2012,Internet Users (%),11.817620
8,2012,Mobile Subscriptions,68.826124
9,2013,Fixed Broadband,1.083984


### 4.1 Trends and Patterns

The analysis shows a clear overall improvement in digital connectivity across the 54 African countries between 2010 and 2024.

- **Internet Users:** Average internet usage increased from 9.05% in 2010 to 44.89% in 2024, showing substantial growth in internet adoption.
- **Mobile Subscriptions:** Average mobile cellular subscriptions increased from 54.74 to 107.25 subscriptions per 100 people. This remained the strongest connectivity channel throughout the period.
- **Fixed Broadband:** Average fixed broadband subscriptions increased from 0.72 to 4.33 per 100 people. Although broadband adoption improved, it remained considerably lower than mobile connectivity.

Overall, the trend indicates that digital connectivity expanded significantly, with mobile networks playing a particularly important role in connecting people across Africa.

In [261]:
relationship_data = long_data.pivot_table(
    index=["Country Name", "Year"],
    columns="Indicator",
    values="Value"
).reset_index()

relationship_data.head(10)

Indicator,Country Name,Year,Fixed Broadband,Internet Users (%),Mobile Subscriptions
0,Algeria,2010,2.486996,12.500000,90.582379
1,Algeria,2011,2.657621,14.900000,96.511295
2,Algeria,2012,3.067372,18.200001,99.685325
3,Algeria,2013,3.332104,22.500000,102.871008
4,Algeria,2014,4.080323,29.500000,110.440351
5,Algeria,2015,5.670601,38.200001,108.016371
6,Algeria,2016,7.000040,42.945499,115.154200
7,Algeria,2017,7.700458,47.691101,109.969863
8,Algeria,2018,7.215668,49.038468,110.938067
9,Algeria,2019,8.275266,55.790699,104.922068


In [262]:
correlation = relationship_data[
    ["Fixed Broadband", "Internet Users (%)", "Mobile Subscriptions"]
].corr()

correlation

Indicator,Fixed Broadband,Internet Users (%),Mobile Subscriptions
Indicator,,,
Fixed Broadband,1.000000,0.622883,0.471026
Internet Users (%),0.622883,1.000000,0.688180
Mobile Subscriptions,0.471026,0.688180,1.000000


### 4.2 Relationships between Variables

Correlation analysis was used to examine the relationships between the three digital connectivity indicators.

- **Internet Users and Mobile Subscriptions:** The correlation coefficient is **0.688**, indicating a moderate-to-strong positive relationship.
- **Internet Users and Fixed Broadband:** The correlation coefficient is **0.623**, indicating a moderate-to-strong positive relationship.
- **Fixed Broadband and Mobile Subscriptions:** The correlation coefficient is **0.471**, indicating a moderate positive relationship.

Overall, the results show that the digital connectivity indicators are positively related. The strongest relationship is between Internet Users and Mobile Subscriptions, suggesting that greater mobile connectivity is associated with higher internet adoption.

In [263]:
change_over_time = relationship_data.groupby("Year")[
    ["Fixed Broadband", "Internet Users (%)", "Mobile Subscriptions"]
].mean()

change_over_time["Internet Users Change"] = (
    change_over_time["Internet Users (%)"].diff()
)

change_over_time["Fixed Broadband Change"] = (
    change_over_time["Fixed Broadband"].diff()
)

change_over_time["Mobile Subscriptions Change"] = (
    change_over_time["Mobile Subscriptions"].diff()
)

change_over_time

Indicator,Fixed Broadband,Internet Users (%),Mobile Subscriptions,Internet Users Change,Fixed Broadband Change,Mobile Subscriptions Change
Year,,,,,,
2010,0.718634,9.054510,54.740167,NaN,NaN,NaN
2011,0.893669,10.285833,62.467689,1.231323,0.175036,7.727522
2012,0.954845,11.817620,68.826124,1.531787,0.061176,6.358435
2013,1.083984,13.885265,74.942558,2.067645,0.129139,6.116434
2014,1.174591,16.238745,78.860474,2.353480,0.090606,3.917916
2015,1.316175,19.205729,82.750307,2.966984,0.141584,3.889832
2016,1.452667,21.903966,81.163456,2.698237,0.136492,-1.586850
2017,1.616788,25.284564,81.619874,3.380599,0.164121,0.456418
2018,1.819429,27.920757,83.143789,2.636192,0.202641,1.523914


### 4.3 Significant Changes Over Time

The year-to-year analysis reveals several significant changes in digital connectivity between 2010 and 2024.

- **Internet Users:** The largest annual increase occurred in 2020, with an increase of **4.68 percentage points**. This indicates a notable acceleration in internet adoption during this period.
- **Fixed Broadband:** The largest annual increase occurred in 2024, rising by **0.94 subscriptions per 100 people**.
- **Mobile Subscriptions:** The largest annual increase occurred in 2024, increasing by **8.25 subscriptions per 100 people**.
- **Decline in Mobile Subscriptions:** A decrease of **1.59 subscriptions per 100 people** was observed in 2016.

Overall, digital connectivity generally followed an upward trend, although some indicators experienced temporary fluctuations. The strongest recent improvement occurred in 2024, particularly in mobile subscriptions and fixed broadband.

In [264]:
# Calculate overall KPI values for the analysis period

kpis = {
    "Average Internet Users (%)": relationship_data["Internet Users (%)"].mean(),
    "Average Mobile Subscriptions": relationship_data["Mobile Subscriptions"].mean(),
    "Average Fixed Broadband": relationship_data["Fixed Broadband"].mean(),
    "Average Digital Connectivity Score": feature_data["Digital Connectivity Score"].mean()
}

kpis

{'Average Internet Users (%)': np.float64(26.144396621626925),
 'Average Mobile Subscriptions': np.float64(81.94278874555924),
 'Average Fixed Broadband': np.float64(1.8480458279659606),
 'Average Digital Connectivity Score': np.float64(0.2492410880108968)}

### 4.4 Key Performance Indicators (KPIs)

The analysis period produced the following key performance indicators:

- **Average Internet Users:** 26.14% of the population.
- **Average Mobile Subscriptions:** 81.94 subscriptions per 100 people.
- **Average Fixed Broadband:** 1.85 subscriptions per 100 people.
- **Average Digital Connectivity Score:** 0.249.

These KPIs provide a summary of the overall level of digital connectivity across the 54 African countries during the 2010–2024 period. Mobile connectivity shows the highest level of adoption, while fixed broadband remains comparatively low.

In [265]:
# Compare average digital connectivity indicators by country

country_comparison = relationship_data.groupby("Country Name")[
    ["Internet Users (%)", "Mobile Subscriptions", "Fixed Broadband"]
].mean()

country_comparison.sort_values(
    "Internet Users (%)",
    ascending=False
).head(10)

Indicator,Internet Users (%),Mobile Subscriptions,Fixed Broadband
Country Name,,,
Morocco,69.112435,127.917733,4.231995
Seychelles,64.799534,142.385924,18.976334
South Africa,58.048153,148.902642,2.561765
Tunisia,56.532000,122.356738,8.453801
Mauritius,55.288062,138.265886,18.804033
Cabo Verde,53.895271,106.656074,4.538694
"Egypt, Arab Rep.",48.663202,93.083864,6.013233
Gabon,47.884347,130.589676,1.485739
Libya,46.691160,151.761739,2.817052


In [266]:
country_comparison.sort_values(
    "Fixed Broadband",
    ascending=False
).head(10)

Indicator,Internet Users (%),Mobile Subscriptions,Fixed Broadband
Country Name,,,
Seychelles,64.799534,142.385924,18.976334
Mauritius,55.288062,138.265886,18.804033
Tunisia,56.532000,122.356738,8.453801
Algeria,46.356518,106.161021,7.006830
"Egypt, Arab Rep.",48.663202,93.083864,6.013233
Cabo Verde,53.895271,106.656074,4.538694
Morocco,69.112435,127.917733,4.231995
Libya,46.691160,151.761739,2.817052
Botswana,39.716640,154.859616,2.797999


### 4.5 Meaningful Comparisons and Insights

A country-level comparison was conducted to identify differences in digital connectivity across the 54 African countries.

The analysis shows that **Morocco** has the highest average Internet Users rate at **69.11%**, followed by **Seychelles (64.80%)** and **South Africa (58.05%)**.

For Fixed Broadband, **Seychelles** has the highest average level at **18.98 subscriptions per 100 people**, followed closely by **Mauritius (18.80)** and **Tunisia (8.45)**.

An important finding is that higher internet usage does not necessarily correspond to higher fixed broadband adoption. For example, Morocco has the highest Internet Users rate but relatively lower fixed broadband adoption compared with Seychelles and Mauritius.

This suggests that internet access in many African countries may rely heavily on **mobile connectivity rather than fixed broadband infrastructure**.

Overall, the comparison highlights significant differences in digital connectivity between countries and suggests that expanding broadband infrastructure could help improve digital inclusion in countries where internet adoption is still relatively low.

# Step 6: Insights and Recommendations

## Key Insights

### 1. What does the data reveal?
The analysis shows a clear improvement in digital connectivity across the selected African countries between 2010 and 2024. Average internet usage increased from 9.05% in 2010 to 44.89% in 2024, indicating significant growth in internet adoption.

### 2. What trends or patterns were discovered?
Internet usage, mobile subscriptions, and fixed broadband generally increased over the period 2010–2024. Mobile subscriptions remained much higher than fixed broadband, showing that mobile connectivity is the dominant form of digital access.

### 3. What are the most important findings?
Mobile subscriptions increased from an average of 54.74 per 100 people in 2010 to 107.25 per 100 people in 2024. Fixed broadband also increased, but remained relatively low, rising from 0.72 to 4.33 per 100 people.

There were also significant differences between countries. For example, Morocco recorded an average internet usage of 69.11%, followed by Seychelles at 64.80%, South Africa at 58.05%, and Tunisia at 56.53%.

### 4. What relationships were identified?
The correlation analysis showed positive relationships between the connectivity indicators. Internet Users and Mobile Subscriptions had the strongest relationship, with a correlation of 0.69. Internet Users and Fixed Broadband had a correlation of 0.62.

These results suggest that improvements in mobile and broadband connectivity are associated with higher levels of internet adoption.

---

## Recommendations

### 1. What actions should stakeholders take?
Governments and telecommunications providers should invest in affordable, reliable, and high-speed broadband infrastructure, particularly in underserved and rural areas.

### 2. What opportunities or risks were identified?
The continued growth in internet adoption presents opportunities for digital education, e-commerce, online services, and digital employment.

However, differences in connectivity between countries and the relatively low level of fixed broadband indicate a continuing digital divide that should be addressed.

### 3. How can the findings support decision-making?
The findings can help policymakers and telecommunications providers identify countries and areas that require greater investment in digital infrastructure. Internet usage, mobile subscriptions, fixed broadband, and the Digital Connectivity Score can be used as key indicators when planning digital development strategies.

## Overall Conclusion

Overall, the analysis shows that digital connectivity in Africa improved considerably between 2010 and 2024. Mobile connectivity remains the dominant form of access, while fixed broadband adoption remains comparatively low. Continued investment in affordable internet services, broadband infrastructure, rural connectivity, and digital literacy can help reduce the digital divide and support inclusive digital development.